src/core/HibpOffline.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ *  Copyright (C) 2019 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "HibpOffline.h"
+
+#include <QCryptographicHash>
+#include <QMultiHash>
+
+#include "core/Group.h"
+#include "core/Tools.h"
+
+namespace
+{
+    const std::size_t SHA1_BYTES = 20;
+
+    enum class ParseResult
+    {
+        Ok,
+        Eof,
+        Error
+    };
+
+    ParseResult parseHibpLine(QIODevice& input, QByteArray& sha1, std::size_t& numPwns)
+    {
+        char hexSha1[SHA1_BYTES * 2];
```

Comment:
Do not use raw char unless ABSOLUTELY necessary. Refactor this PR to use QChar, QString, and/or QByteArray in all places.

Reply:
After reading your above responses... optimization is not an excuse to use fixed array sizes. This operation only runs when the user requests it and does not require micro-optimizations to save an extra 50 milliseconds.

Reply:
I looked into using the [`QIODevice::read()`](https://doc.qt.io/qt-5/qiodevice.html#read-1) overload that returns a `QByteArray`. However, as its docs say:

> This function has no way of reporting errors

Therefore, I can no longer distinguish between EOF and I/O error, which is a problem.

If we want to just avoid `char` at all costs, I could of course do:

```c++
QStringArray hexSha1(SHA1_BYTES * 2, '\0');
const qint64 rc = input.read(hexSha1.data(), hexSha1.size());
```

I fail to see how that is any better, since it's just as vulnerable to a buffer overrun if I accidentally pass in the wrong size as the second parameter of the `input.read()` call (which I assume is your concern). But I can do it if required -- let me know.

Reply:
It is fine, after chatting with the other team members we can proceed with it as-is.